csrc/preseg_passes/fmin_fmax_promotion.cpp

```diff
@@ -0,0 +1,298 @@
+  }
+
+  // The only remaining cases are combinations of DEFAULT,
+  // BAD_BROADCAST, and BAD_BROADCAST_DEFAULT.
+  return IterDomainStatus::BAD_BROADCAST_DEFAULT;
+}
+
+bool StatusIsBad(IterDomainStatus status) {
+  return status == IterDomainStatus::BAD_REDUCE ||
+      status == IterDomainStatus::BAD_BROADCAST ||
+      status == IterDomainStatus::BAD_BROADCAST_DEFAULT;
+}
+
+bool AnyBadInputs(Expr* expr, IterStatusMap& iterMap) {
+  for (auto input : expr->inputs()) {
+    if (auto* in_tv = dynamic_cast<TensorView*>(input)) {
+      for (IterDomain* id : in_tv->getLogicalDomain()) {
+        IterDomainStatus status = iterMap[id];
+        if (StatusIsBad(status)) {
+          return true;
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+// Once we identify a target reduction, we perform a downward pass starting from
+// the target's direct input. The pass propagates IterDomainStatus information.
+// At the end, we check all output TV's for bad statuses. If at any point we
+// encounter a node we don't know how to propagate information through, we treat
+// it like to a graph output and fail if it has any incoming bad statuses.
+bool AnalyzeMinMaxOp(ReductionOp* targetRop) {
+  Fusion* fusion = targetRop->fusion();
+
+  FusionGuard fg(fusion);
+  ComputeAtLogicalDomainMap logical_map;
+  logical_map.build(true);
+
+  IterStatusMap iterMap;
+
+  auto* in_tv = targetRop->input(0)->as<TensorView>();
+  for (IterDomain* in_id : in_tv->getLogicalDomain()) {
+    iterMap[in_id] = IterDomainStatus::DEFAULT;
+  }
+
+  auto* out_tv = targetRop->output(0)->as<TensorView>();
+  for (IterDomain* out_id : out_tv->getLogicalDomain()) {
+    if (out_id->isReduction()) {
+      iterMap[out_id] = IterDomainStatus::BAD_REDUCE;
+    } else {
+      iterMap[out_id] = IterDomainStatus::DEFAULT;
+    }
+  }
+
+  auto traversal =
+      StmtSort::getExprsBetween({targetRop->input(0)}, fusion->outputs());
+  for (Expr* expr : traversal) {
+    std::string opName = expr->getOpString();
+
+    if (expr == targetRop) {
+      // Skip the target rop. We already marked its status.
+      continue;
+    }
+
+    bool anyBadInputs = AnyBadInputs(expr, iterMap);
+
+    auto* out_tv = dynamic_cast<TensorView*>(expr->output(0));
+
+    if (!out_tv) {
+      if (anyBadInputs) {
+        return false;
+      } else {
+        continue;
+      }
+    }
+
+    if (expr->isA<UnaryOp>() || expr->isA<ReductionOp>() ||
+        expr->isA<BroadcastOp>()) {
+      auto in_tv = expr->input(0)->as<TensorView>();
+      auto p2c = logical_map.mapBestEffort(
+          in_tv->domain(),
+          in_tv->getLogicalDomain(),
+          out_tv->domain(),
+          out_tv->getLogicalDomain());
+
+      for (IterDomain* in_id : in_tv->getLogicalDomain()) {
+        IterDomainStatus status = iterMap[in_id];
+        auto out_id = p2c[in_id];
+
+        if (out_id) {
+          if (out_id->isReduction()) {
+            if (status == IterDomainStatus::BAD_BROADCAST) {
```

Comment:
So, here we have a broadcast input ID that is reduced. This isn't supposed to happen. Instead, it should be squeezed. Did you encounter anything particular? We did allow reduction of broadcast IDs long time ago, and we realized it's a bad idea and change it to squeezing. There may be some places where we may be still doing so.